File: 10/perl/10.pl
#!/usr/bin/perl

use strict;
use warnings;

use integer;

my $start;

BEGIN
{
  $start= time;
}

END
{
  print "Time: " . (time - $start) . "\n";
}

my $sum = 0;

for(1..2000000)
{
  if(isPrime($_))
  {
    $sum += $_;
  }
}
print "Answer: $sum\n";

sub isPrime
{
  my $n = shift;

  my $root = sqrt $n;
  $root++;

  return 0 if $n < 2;
  return 1 if $n == 2;
  return 0 if $n % 2 == 0;

  my $i = 3;

  while($i < $root)
  {
    if($n % $i == 0)
    {
      return 0;
    }
    $i+=2;
  }
  return 1;
}

File: 10/cpp/10.cpp
#include <cstdio>
#include <cmath>
#include <ctime>
#include <string.h>

#define MAX 2000000

bool isPrime(unsigned long long);

int main(void) 
{
  std::time_t start = std::time(0);
  unsigned long long sum = 2;

  for(unsigned long long j = 1; j < MAX; j+=2)
  {
    if(isPrime(j))
    {
      sum += j;
    }
  }

  printf("Answer: %lld\n", sum);
  std::time_t end = std::time(0);
  printf("Time: %d\n", end - start);
  return 0; 
}

bool isPrime(unsigned long long n)
{
  unsigned long long root = std::sqrt(n);
  ++root;

  if(n < 2)
  {
    return false;
  }
  if (n == 2 || n == 3)
  {
    return true;
  }
  if (n % 2 == 0)
  {
    return true;
  }

  for (unsigned long long i = 3; i < root; i+=2)
  {
    if(n % i == 0)
    {
      return false;
    }
  }
  return true;
}
